source/base/parameter_handler.cc

```diff
@@ -2379,6 +2402,37 @@ ParameterHandler::operator==(const ParameterHandler &prm2) const
 }
 
 
+std::set<std::string>
+ParameterHandler::get_entries_wrongly_not_set() const
+{
+  std::set<std::string> entries_wrongly_not_set;
+
+  for (auto it = entries_set_status.begin(); it != entries_set_status.end();
+       ++it)
+    if (it->second.first == true && it->second.second == false)
+      entries_wrongly_not_set.insert(it->first);
+
+  return entries_wrongly_not_set;
+}
+
+void
+ParameterHandler::assert_all_entries_have_been_set() const
+{
+  std::set<std::string> entries_wrongly_not_set =
+    this->get_entries_wrongly_not_set();
+
+  AssertThrow(
+    entries_wrongly_not_set.size() == 0,
```

Comment:
The fact that you can write the assertion in what is essentially a statement function suggests that it isn't very useful. I still think that we shouldn't have this function. One reason is that right now, you really only print one of the missing parameters, but that will lead to a long sequence of trial and error in practice where every time you call a program, you get told about yet another parameter that is missing. But you never get to see all at once. I don't think that's a useful design.

Reply:
Maybe we mix up different aspects here. Is the problem that it is a public function or would the problem be the same in case of a private one? The code provides the flexibility to do what you want. One can call the parse_input() function with assert_all_entries_have_been_set = false (=default), and subsequently call get_entries_wrongly_not_set() where the user can then print the whole list. For me it is ok if we print the whole list by default. The sentence with "trial and error" sounds a bit negative to me given that I want to improve the ParameterHandler in a way to avoid **errors**.

Reply:
The way I approach this is like this:

* Every addition of a function has a cost. It has to be maintained, its documentation has to be kept up to date, it expands the number of functions users have to look through to find what they are looking for, it tempts users to use this function even in cases where it doesn't quite do what they need and then prompts them to write work-arounds, etc. So adding a function is not free.

* This has to be balanced with the question of whether the function (or a variation of it) is *useful*. 

I claim that that the `assert` function is *not actually useful enough* to justify the cost. That's because what it does is trigger an assertion that only shows one parameter that isn't set. But if multiple parameters aren't set in an input file, you still only get to see one of them in the error message. So if you call your application with an input file that misses multiple parameters, you get to see one missing parameter; you fix it; you call the application again; you get another error; you fix that; call the application again; etc. 

A better error message would probably list all critical-but-missing parameters at once. I suppose that I could accept such a function -- it's useful, and I understand your use case and why you want to have it. It's mainly that we don't have these kinds of functions in the library very much, but maybe one can argue that such a function would really not be all that different from having assertions that the input to a `ParameterHandler` must be a well-formed file that follows a certain syntax -- in your case, the syntax just happens to include a clause that certain elements need to be present.

Reply:
For example, @peterrum wants to have the Assert in the other ``parse_input_...()`` functions as well (which is more than one), so doesn't this already imply that we need an assert-function. Since the plan is to make the ``parse_input()`` function of this PR a member function, the assert-function could then become a private member function. Would this be a compromise?

Reply:
I disagree with @peterrum on this :-) I don't think the assertion should be in there because there are use cases where people parse several `.prm` files one after the other. One of them isn't going to fill all parameters, and so the assertion would be misplaced there.

Reply:
I understand your point. At the same time it seems to me that this is not how the ParameterHandler is currently designed. For example, if a parameter does not match the predefined pattern, we currently also have to "debug" through the code in a "last man standing" style. I think the code throws after the first parameter has been found that does not match its pattern. 

A second point is that, for my understanding, the cost of a function is always justified if it avoids code duplication. I would say that in an application code you need around 10-20 lines to implement a routine that gets the set from the ParameterHandler, then outputs the missing parameters, and finally throws an exception. Every application code will have to do that, in reality some application codes will probably implement it more than once. My point of view is probably more application-oriented, but wouldn't it make sense that this functionality is already part of deal.II.

Reply:
@bangerth What I suggested yesterday is to introduce to the `parse_input_*`-functions an additional optional parameter (maybe with the name) `assert_all_parameters_have_been_set`. The reason why I suggested it is that I have the feeling that in many cases - especially if you work with a single parameter file - both undefined entries and parameters not set are unwanted behavior and might be as such asserted.

The usage could be in the case that only one input file is used:
```cpp
prm.parse_input (file, ..., true);
```
and in the case of multiple input files:
```cpp
prm.parse_input (file1, ..., false);
prm.parse_input (file2, ..., false);
prm.parse_input (file3, ..., false);
prm.assert_that_parameters_have_been_set ();
```

A function whose only purpose is to assert is an option. An other option would be to introduce a new exception type:
```cpp
Assert(prm.are_all_parameters_set(),
        Exception::NotSetParameter(prm.get_any_not_set_parameter()) );
```

It is some copy and paste, but not that much. What do you think @nfehn?

Reply:
Yes, this is what I suggested (the last option is not my favorite due to DRY, see above). I mean, no one is forced to run into the assert. We can set the default value to false so that nothing happens. Unfortunately, we need a default value because the ``parse_input()`` functions already have something like ``skip_undefined`` with default value and we now can not change this backwardly. Theoretically, the best design would certainly be to give the user all the freedom and, at the same time, not fool the user by default values that lead to erroneous results because the user has not been aware of them.